### ISSUE TITLE
Question about Reverse Huber Loss

### ISSUE THREAD

**Issue by leonardoaraujosantos** [iro-cp/FCRN-DepthPrediction#16]
Hi @iro-cp I'm taking a look on the paper and got confused on the section 3.2 (Loss Function)

![revloss](https://cloud.githubusercontent.com/assets/898383/24552491/944e008c-161e-11e7-91df-e1ff43a68624.png)

Where c
![cvalue](https://cloud.githubusercontent.com/assets/898383/24552961/686c5d86-1620-11e7-9b46-76f79b9bc612.png)

## Questions
* c Will be a scalar related to 20% (1/5) max error on the batch correct?
* When I need to select between L1 and Quadratic loss. More specifically the  condition |x| < c. Does the |x| need to be summed over all it's elements in order to compare with this scalar number "c"?

## Example
Considering the vector |x| with a batch of 3 elements as ...
![l1x](https://cloud.githubusercontent.com/assets/898383/24552737/838e627c-161f-11e7-9fc4-d12faedae201.png)

My worst error will be 6, so c will be 1/5*6
c = 1.2

So summing all elements
![suml1](https://cloud.githubusercontent.com/assets/898383/24552808/dcded884-161f-11e7-8b8d-8e344f6689ec.png)

|x| will be 9, then on this case the quadratic part of the reversed Huber would be selected

Or we are completely wrong and we should calculate B(x) for each element of |x| on the batch and then sum all those losses and divide by the batch size?

**Comment by iro-cp** [comment]
Hey, 

To answer your first question, the way we calculate c is correct like this (also in your example). 

The second part of your example should be different though: Instead of summing up elements and then deciding whether to apply L1 or quadratic Loss, this decision actually happens per element. 
However, the "elements" are now all the image pixels. So every sample in your batch corresponds to an image and every pixel of the image gets penalized by either term depending on whether its difference to the ground truth value is smaller or larger than c. 

Given the differences in your example, you would apply L1 loss to the first element, and quadratic on the other two. Then sum up.

**Comment by leonardoaraujosantos** [comment]
Hi @iro-cp thanks for the help.

**Comment by Cuky88** [comment]
Can you please tell me if my implementation of the berHu loss is correct:
```
def berhu_loss(labels, predictions, scope=None):
    if labels is None:
        raise ValueError("labels must not be None.")
    if predictions is None:
        raise ValueError("predictions must not be None.")
    with tf.name_scope(scope, "berhu_loss",
                        (predictions, labels)) as scope:
        predictions = tf.to_float(predictions)
        labels = tf.to_float(labels)

        # Make sure shape do match
        predictions.get_shape().assert_is_compatible_with(labels.get_shape())

        # Get absolute error for each pixel in batch 
        abs_error = tf.abs(tf.subtract(predictions, labels), name='abs_error')

        # Calculate threshold c from max error
        c = 0.2 * tf.reduce_max(abs_error)

        # if, then, else
        berHu_loss = tf.where(abs_error <= c,   
                       abs_error, 
                      (tf.square(abs_error) + tf.square(c))/(2*c))
                
        loss = tf.reduce_sum(berHu_loss)

    return loss
```
